### ISSUE TITLE
Impossible to set data for a plot that had no initial data 

### ISSUE THREAD

**Issue by p-himik** [bokeh/bokeh#6639]
Consider this snippet (intended for `bokeh serve`):

```python
from bokeh.layouts import row
from bokeh.plotting import figure, curdoc
from bokeh.models import ColumnDataSource, Button

s1 = ColumnDataSource(data=dict(x=range(10), y=range(10)))
f1 = figure()
f1.circle(x='x', y='y', source=s1)

s2 = ColumnDataSource(data=dict(x=[], y=[]))
f2 = figure()
f2.circle(x='x', y='y', source=s2)

b = Button(label='Update')

def update_s2():
    s2.data = dict(x=range(10), y=range(10))

b.on_click(update_s2)

curdoc().add_root(row(f1, f2, b))
```
`f1` works just fine - the ranges are computed correctly, the figure is shown without any warnings.
But since the default ranges for `f2` are `DataRange1d` and there's no initial data in `s2`, you'll see `[bokeh] could not set initial ranges` in the console. Pressing "Update" button will correctly update the second plot only if there was no interaction. If there was interaction, however, the console will be flooded with `[bokeh] could not set initial ranges` messages and pressing "Update" will do nothing.

Now, if we set ranges of `f2` to be `DataRange1d(start=0, end=1)` to get rid of the console warning about initial ranges, pressing "Update" will indeed change the plot - but the ranges will stay the same since providing explicit start and end to `DataRange1d` effectively turns it into `Range1d`.

It's pretty simple to solve this issue on a user side by using `CustomJS` callback to update `f2` ranges whenever `s2` updates. But it becomes really tiresome when there are multiple glyphs with different sources.
It also should be possible to fill empty data sources with a dummy data and make all glyphs invisible before updating the data. But it is not the best approach as well - a user will have to manually change glyphs' visibility.

At this point, I don't know any other solution and frankly, I have no idea how to fix it on the Bokeh side. Maybe implement implicit initial ranges or just somehow go around the error with no initial ranges set?

**Comment by bryevdv** [comment]
This seems to be working fine for me on `0.12.10dev` master. Update always works, regardless of any previous panning, zooming, etc. Please re-open if I have misunderstood the report.

**Comment by p-himik** [comment]
@bryevdv Maybe I did something wrong, but I just pulled master, installed all necessary packages, and built Bokeh again:
```
 % python -m bokeh info                                                                  
Python version      :  3.5.4 | packaged by conda-forge | (default, Aug 10 2017, 01:38:41)
IPython version     :  6.1.0
Bokeh version       :  0.12.10dev3-2-g7e4f0e3
BokehJS static path :  /home/p-himik/dev/git/bokeh/bokeh/server/static
node.js version     :  v6.11.4
npm version         :  3.10.10
```

And the example above still doesn't work. After I try to pan the second plot before updating it, I can see a huge number of `[bokeh] could not set initial ranges` messages in the console and "Update" does nothing except causing another message about initial ranges to appear.

**Comment by p-himik** [comment]
@bryevdv If you link the plots e.g. by calling `f2.x_range = f1.x_range`, it won't be possible to update the second plot at all - no matter whether there was some interaction or not.

**Comment by psalios** [comment]
@p-himik @bryevdv I have the code snippet above with the current version `0.12.14` and it works.

**Comment by chesucr** [comment]
Yes, for me it works for `1.0.0dev8`, the data of the second plot is updated as well, but I am still getting the warning:

    [bokeh] could not set initial ranges

If several sources and renderers are being used, the [`renderers` attribute](https://bokeh.pydata.org/en/latest/docs/reference/models/ranges.html#bokeh.models.ranges.DataRange.renderers) can be set as well to avoid the warning.

**Comment by bryevdv** [comment]
Noting this is still an issue on  `1.1.1dev1` and also noting as a reminder that panning the empty plot is required to expose the issue.

**Comment by bryevdv** [comment]
Still an issue in 3.0 (panning required) Here is an updated MRE:
```python
from bokeh.layouts import row
from bokeh.plotting import figure, curdoc
from bokeh.models import ColumnDataSource, Button

s1 = ColumnDataSource(data=dict(x=list(range(10)), y=list(range(10))))
f1 = figure()
f1.circle(x='x', y='y', source=s1)

s2 = ColumnDataSource(data=dict(x=[], y=[]))
f2 = figure()
f2.circle(x='x', y='y', source=s2)

b = Button(label='Update')

def update_s2():
    s2.data = dict(x=list(range(10)), y=list(range(10)))

b.on_click(update_s2)

curdoc().add_root(row(f1, f2, b))
```

**Comment by bryevdv** [comment]
This seems to have been fixed at some point, result after hitting update:

<img width="1273" alt="Image" src="https://github.com/user-attachments/assets/223b09f4-81a3-46b2-ab1f-1c16607a4997" />

@mattpap can you confirm and/or speculate about when it might have been resolved?

**Comment by mattpap** [comment]
> can you confirm and/or speculate about when it might have been resolved?

No. I would suggest a bisection to figure that out and a regression test to seal this behavior in.

**Comment by bryevdv** [comment]
I would generally agree but eight years is a long time and I really doubt I could even get 0.12.6 to build at this point. I'm just going to close and we can deal with new issues (or re-occurrences) as they are reported.

**Comment by bryevdv** [comment]
Well, I guess per my previous comment we'd only need to go back to 3.0 which I can build. If I can track something down very quickly, I will update this, otherwise I will just let it lie. 

Edit: eh, I guess our history is unhappy in some way
```
dev312 ❯ git bisect bad 3.0.0
Some good revs are not ancestors of the bad rev.
git bisect cannot work properly in this case.
Maybe you mistook good and bad revs?
```
I don't have cycles to try and pursue this any more unfortunately.
